Extend ocamlc -output-complete-obj to build .exe

nojb opened the pull request:
This PR extends `ocamlc -output-complete-obj` to build self-contained `.exe`'s for bytecode programs containing both the runtime and any linked C stubs.

The difference with `-custom` is that here we embed the C bytecode directly in the generated C code file so the resulting binary is more robust than the one produced by `-custom` (eg it can be stripped). 

Note that contrary to binaries produced by `-custom` the resulting executable cannot be used with tools that need access to the bytecode/symbol table information (eg `ocamldebug`).

The idea is to deprecate `-custom` in a follow-up PR.

The code in this PR is based on a patch written by @glondu that modified `-custom` to obtain essentially the same effect, but it was decided to leave `-custom` alone because of backwards-compatibility issues (see previous version of this issue description for more information).

nojb committed:
Rebase Stéphane Glondu's patch

alainfrisch commented:
This new implementation requires a native toolchain, contrary to the current one.  This might exclude some use cases.

glondu commented:
> This new implementation requires a native toolchain, contrary to the current one.

What do you mean by "native"? Isn't a toolchain already needed to build the custom runtime, no matter how you "link" it with the bytecode?

nojb commented:
> This new implementation requires a native toolchain, contrary to the current one. This might exclude some use cases.

As far as I can see, both implementations make the same call to the native toolchain using `Ccomp.call_linker` (via `Bytelink.build_custom_runtime`). Am I missing something?

nojb commented:
One issue is that thanks to putting the bytecode TOC at the end of the bytecode, tools that work with bytecode could work equally well with `-custom` and usual bytecode programs, as they can jump over the custom runtime code seamlessly.

After this PR, this is no longer the case. For example, `ocamldebug`, `ocamlobjinfo` and `cmpbyt` (used by `ocamltest`) no longer work with `-custom` binaries.

ghost commented:
If we want to keep this feature, one idea would be to put a marker before the bytecode so that tools can extract it.

alainfrisch commented:
Yep, sorry, I was thinking of the mode that allows building a custom runtime, and then concatenating some bytecode program to it (without the need to have a C compiler/linker at that stage).

glondu commented:
I was just reminded [this bug](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=678577). It might be useful to read in the context of this PR.

glondu commented:
Another [Debian bug](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=627761) related to the new behaviour of -custom. I remember now why I made the behaviour Debian-specific...

glondu commented:
I went through test failures.

Many of them compare the outputs of ocamlc.byte and ocamlc.opt using a custom tool called `cmpbyt` that parses bytecodes and ignores debug sections. This obviously fails with the new `-custom` behaviour. But it turns out that the outputs are identical, so adding a whole-file comparison to `cmpbyt` as done in https://github.com/glondu/ocaml/commit/bf9434d23c4ccf7fb780ee8a5058ae0adb31d5b0 fixes these tests.

Another test fails because of a missing C include: https://github.com/glondu/ocaml/commit/ad34e4c1d4d2b875c50afae0991226f170a41bdb.

The last three fail because they try to link in a custom runtime C code that defines a `main()` function. I didn't even know that was possible, and wonder what the semantics could be.

gasche commented:
I find some of the limitations mentioned on the Debian bugtracker (having `ocamlrun` and `ocamldebug` stopping working on `-custom` bytecode executables) are rather severe (as mentioned in [this message](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=627761#5), a priori fixing stripping of bytecode executables may not be an aspect that users care about as much). Does making this change upstream allow to consider fixing these limitations? Maybe ocamlrun/ocamldebug and other bytecode-consuming programs could be changed to work correctly with that new representation?

glondu commented:
> I find some of the limitations mentioned on the Debian bugtracker (having `ocamlrun` and `ocamldebug` stopping working on `-custom` bytecode executables) are rather severe

I consider the ability to run `ocamlrun` and `ocamldebug` on `-custom` bytecode executables with the old behaviour a fortunate anomaly: it means `-custom` was not needed in the first place.

gasche commented:
Yes, but this anomaly may still be relied upon by an unknown amount of users whose workflow would break with the change -- which does give me cold feet.

glondu commented:
The workaround in this specific case is easy.

xavierleroy commented:
ocamldebug is perfectly able to work with "true" custom bytecode executables, containing a non-standard runtime:
```
~/tmp$ cat foo.ml
let _ =
Printf.printf "%f\n" (Unix.gettimeofday())
~/tmp$ ocamlc -custom -g -o foo.exe unix.cma foo.ml
~/tmp$ ocamldebug foo.exe
	OCaml Debugger version 4.07.1

(ocd) go 100
Loading program... done.
Time: 100 - pc: 140316 - module CamlinternalFormat
272   <|b|>buffer_check_size buf str_len;
(ocd) 
```
Don't break this just to please Debian's obsession with stripped binaries.

ghost commented:
It works with `ocamldebug` but not with `ocamlrun`. Looking at htop, I'm guessing that `ocamldebug` simply starts the program and instructs the runtime via an environment variable to run in debug mode and connect to the `ocamldebug` process via a socket. Is that correct?

In that case, whatever we do here is unlikely to break `ocamldebug` working with custom built bytecode executables.

xavierleroy commented:
> In that case, whatever we do here is unlikely to break ocamldebug working with custom built bytecode executables.

ocamldebug needs to 1- start the program instructing the VM to connect to the debugging socket, and 2- read symbol table and debug information from the bytecode executable.  Your current proposal does not break 1, but breaks 2 as far as I understand.  Please don't.

ghost commented:
I suggest to put a marker at the beginning of the symbol table and debug information. Then `ocamldebug` could scan the binary looking for these markers.

xavierleroy commented:
> I suggest to put a marker at the beginning of the symbol table and debug information.

Apologies if I sound negative, but I think this is a bigger, uglier hack than the one it replaces.

glondu commented:
I agree with @xavierleroy here.

I think we can cleanly locate the ELF symbols corresponding to the OCaml [symbol table and debug information] using libbfd or such. Or have the runtime transmit [them] over the debugging socket at startup.

ghost commented:
Both of @glondu ideas seem good to me. The ELF symbol idea seems more generic if we want other tools to be able to easily extract these values. Plus it's the method already used by `objinfo_helper`.

xavierleroy commented:
MacOS and Windows don't use ELF.

ghost commented:
I had a feeling that wouldn't be portable... I thought libbfd might but I wasn't sure. That's why I suggested my ugly hack :see_no_evil: 

Transmitting the info over the debugging socket should be fine, right?

alainfrisch commented:
Another thing that works today is to build a `-custom` executable, but pass it to an explicit custom runtime (built by `ocamlc -make-runtime`).  The number of use cases that could be broken by the suggested change makes me a bit nervous about it.

IIUC, the proposed new behavior is very similar to using `-output-complete-obj foo.{so,dll}`.  What about extending `-output-complete-obj` to support producing programs instead of shared libraries (or introducing a new flag with the same effect), and keep `-custom` unchanged?

dra27 commented:
Rather than scanning binaries at all, given that we already have a special way of invoking the *runtime* to attach to a debugger, can’t we embed the bytecode as here but also have special ways of invoking the custom runtime binary which cause it to emit its bytecode (for `ocamlrun`) and symbol and debugging info (for `ocamldebug`) - wouldn’t the only challenge then be having both `ocamlrun` and `ocamldebug` spot the difference between a bytecode image and a custom runtime image?

dra27 commented:
(A similar mechanism might allow the case when `header.c` is used instead of shebang to survive stripping, not that that affects Debian)

ghost commented:
> Another thing that works today is to build a -custom executable, but pass it to an explicit custom runtime (built by ocamlc -make-runtime). The number of use cases that could be broken by the suggested change makes me a bit nervous about it.

@alainfrisch do you expect anyone to rely on this behavior? It's difficult to imagine a use case for constructing both a custom executable and a custom runtime. It seems more natural to choose one or the other.

alainfrisch commented:
One could imagine building a custom bytecode executable, but sometimes execute it with a custom runner built with the debug runtime (or another kind of runtime variant, e.g. with extra hooks).  Another use case would be creating programs which could also be executed either in a stand-alone way or through a native host program (that would include the OCaml runtime), without having to create a new process.

ghost commented:
So, just to sum up all these discussions:

- the current `-custom` has been considered a dirty hack for a while
- once upon a time, there was a plan to get rid of it
- it never happened and we currently rely on it in Dune for valid reasons
- the fact that Dune relies on `-custom` makes Debian sad
- we have a clean re-implementation of `-custom` that makes Debian happy, however this implementation looses a few features compared to the current `-custom`
- there are valid cases for such features, though we don't know if they exist in the wild
- we have a few ideas to recover such features with the new implementation

Given all this, and in the interest of not over-engineering things unless it is really necessary, I suggest to follow @alainfrisch's suggestion; we introduce a new command line argument to select the new cleaner method and leave `-custom` unchanged. Executables produced with the new method will not support debugging via `ocamldebug` and will not support being executed with a custom `ocamlrun`. If someone ever complain about that, we'll re-evaluate the various solutions proposed here.

Then, if someone is motivated we can resume the original plan to get rid of `-custom`.

Khady commented:
> Executables produced with the new method will not support debugging via ocamldebug and will not support being executed with a custom ocamlrun.

Sorry to step in the discussion without a valuable technical input, but I have a question. Does it mean that executables generated by dune won’t support debuting via ocamldebug?

dra27 commented:
I have a feeling that the proposed new flag results in Dune producing “bytecode” images where when a user does discover that it doesn’t work with `ocamlrun` or the debugger will cause OCaml to be blamed initially, not Dune 🙂

ghost commented:
Pure bytecode programs, i.e. with extension .bc, would still support ocamldebug. The two situation where `ocamldebug` is currently supported and wouldn't after is:

- the `.exe` program when the executable stanza as `(modes byte)`
- the `.exe` program on bytecode-only architectures

Would that be an issue? This feature would be easy to recover using @glondu's method of sendong the necessary information through the debugging socket.

dra27 commented:
Or, to put it another way, it feels like we have a `-custom` option which has bad points and we now propose a new option which has different bad points instead.

ghost commented:
I see. TBH, I feel like this `ocamlrun` buisiness is far-fetched.

ghost commented:
We just had a chat with @dra27 about this. We both think that the best is to:

- switch `-custom` to the new method, as proposed by this PR
- support `ocamldebug` via @glondu's method
- forget about running custom built executables with `ocamlrun`

alainfrisch commented:
Aren't people using `-custom` with js_of_ocaml, to make the compiler happy with custom primitives (passing dummy stub C implementations)?  `-no-check-prims` has been introduced for that purpose, but I guess people previously relied on `-custom` and some projects might not have switched.

Perhaps there are even some cases where people actually want to produce both an executable (with -custom, using real C primitives) and also pass it to `js_of_ocaml`.

All those concerns are perhaps theoretical, but it's hard to be convinced that by breaking the assumption that the bytecode can be extracted from executables built with `-custom`, we are not breaking existing use cases in the wild.  I'm not necessarily against taking the risk, but the decision should not be taken lightly.

dra27 commented:
I can see the point that if this is done and the behaviour subsequently found in the wild, then we break someone's albeit slightly niche use of OCaml on the altar of a philosophy, so as @xavierleroy said last week, perhaps we should step back to ask why again! <sigh>

At the moment, the exec_magic_number is expected to be at the end of the file. Might an option be to combine various ideas previously and simply relax the loading as being a scan from the end of the file to search for the exec magic number? That scan only comes into play in the corner cases (if I may place the use of the debugger as being a "corner" case) - `#!` executables are unaffected, as are the new-style `-custom` executables themselves, it's only a specific runtime invoked on a `-custom` executable or the debugger which might have to scan further into the file to find the bytecode image.

So we could as previously suggested use appropriate compiler pragmas to place the data in a custom section (ELF, Mach-O and PE do all support this) and similarly enough linker fu to place that section as near if not at the end of the image as possible (crucially, we want it after any other user data which might accidentally contain the exec_magic_number). `configure` could be responsible for falling back to the old-style implementation for platforms where it happens to be known that the section/linker tricks aren't available or don't work. At this point, any attempts at compressing/encrypting the executable image will change (that's a minor potential concern on Windows, but easily solved by the statement "Don't do this with bytecode!"). That has the benefit of avoiding the runtime having to be built against any binutils libraries, which sounds like an MSVC-nightmare and also avoids more special cases in the runtime to extract the data.

Given the scanning, it might also be worth strengthening the bytecode header to include the total length of the file (I think the length at the moment is only the section table) and possibly also a checksum of the entire image.

ghost commented:
@hhugo, is `-custom` still necessary for `js_of_ocaml`?

> `-no-check-prims` has been introduced for that purpose, but I guess people previously relied on `-custom` and some projects might not have switched.

Why are such projects a concern? If they rely on older versions of the tools, I suppose they are using an older compiler as well.

Personally, I'm concerned about how big this project is becoming. If we start messing around with the various exe formats, it's both going to take longer to implement and longer to debug as well. It seems worth supporting `ocamldebug`, but I still don't see the point of supporting running `-custom` executables with `ocamlrun`.

nojb commented:
One way of sidestepping this discussion is to leave `-custom` alone and instead extend `-output-complete-obj` with an `.exe` target that does what the present PR is doing : generate a binary with the runtime and C libs linked in. The downside is that it becomes one more combination to support, which increases the possibility of bugs, etc.

dra27 commented:
@nojb - it's not just the possibility of bugs, it means there would be two kinds / three formats of bytecode executable to worry about (the kind which works with the `debugger` and `ocamlrun` and the kind which doesn't; traditional `-custom` concatenation format, shebang format, and embedded format).

ghost commented:
We just had a discussion about this with @dra27, @nojb, @glondu and @xavierleroy. In the end, we all agreed that while `-custom` is solving a problem that is still relevant today because of exotic architectures officially supported by Debian, it is not something we expect users to explicitely use.

Given that and in the interest of not multiplying the features supported by the compiler, the plan we agreed on is as follow: we will extend `-output-complete-obj` as @nojb suggested and in parallel we will deprecate `-custom`. Dune will switch from using `-custom` to using `-output-complete-obj` as a fallback on bytecode only architectires. After two releases of the compiler, if the `-custom` deprecation went well, we will delete it entirely just as we did for `-thread/-vmthread`.

nojb committed:
Extend -output-complete-obj to target .exe

nojb commented:
I pushed a commit implementing the extension to `-output-complete-obj`.

nojb commented on bytecomp/bytelink.ml:
```diff
@@ -656,8 +674,13 @@ let link objfiles output_name =
       (fun () ->
-         link_bytecode_as_c tolink c_file;
-         if not (Filename.check_suffix output_name ".c") then begin
+         let with_main = Filename.check_suffix output_name ".exe" in
```
Note that the `.exe` extension is required in the argument to `-o` even on Unix.

diml replied:
Indeed, I guess it might be slightly surprising for pure unix users, but at least it's in line with the current behaviour of `-output-complete-obj`. If that was really an issue in pratice, we could add a way to specify the output format instead of detecting it via the extension.

nojb committed:
Changes

nojb committed:
Add test

nojb commented:
@diml This PR is ready for review. I have updated the PR title and description with relevant information.

ghost commented:
Reviewing it now :)

dbuenzli commented:
> Note that the .exe extension is required in the argument to -o even on Unix.

It seems this PR adds yet another undocumented way of working to the uterly confusing way the `-o` option works in `ocaml`. Why not produce an object if a platform specific object extension is detected and an executable otherwise ?

ghost commented:
What would be the story for filenames that have a dot but no known extension?

diml commented on driver/main.ml:
```diff
@@ -213,7 +213,8 @@ let main () =
             || Filename.check_suffix s Config.ext_dll
-            || Filename.check_suffix s ".c")
+            || Filename.check_suffix s ".c"
+            || Filename.check_suffix s ".exe")
```
The ".exe" extension should be added to the error message bellow

nojb replied:
Thanks, fixed.

ghost commented:
`manual/manual/cmds/intf-c.etex` describes the usage of `-output-obj` and in particular how it behaves depending on the output file extension. That's probably the right place to describe what happens when the output filename ends with `.exe`.

nojb committed:
Add doc

nojb commented:
> `manual/manual/cmds/intf-c.etex` describes the usage of `-output-obj` and in particular how it behaves depending on the output file extension. That's probably the right place to describe what happens when the output filename ends with `.exe`.

Thanks, I had missed that. I pushed a commit with some explanation.

nojb commented:
Currently this option exists only for `ocamlc`. Should we also allow it for `ocamlopt` (as a no-op) ?

nojb commented:
> What would be the story for filenames that have a dot but no known extension?

I think @dbuenzli is suggesting that in this case we would default to producing an executable with that name.

dbuenzli commented:
> I think @dbuenzli is suggesting that in this case we would default to producing an executable with that name.

This would be more in line with the way the toolchain works. It seems a bit odd to attribute special semantics to the `.exe` suffix which is a Windows thing. 

The toolchain consistently allows you to produce executables without extension and then you can simply add `.exe` if you are on Windows. With this I now need to pay attention if my executables are being produced via this path and special case for it which is very annoying.

> Note that contrary to binaries produced by `-custom` the resulting executable cannot be used with tools that need access to the bytecode/symbol table information (eg `ocamldebug`).
> 
> The idea is to deprecate `-custom` in a follow-up PR.

I'm still a bit confused about all this. It seems @xavierleroy "deprecated" `-custom` 10 years ago in a random [debian bug report](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=256900#49) in favor of bytecode+C dynlink. 

However for me the idea of `-custom` was that it would produce the moral equivalent *for bytecode* of what `ocamlopt` does (i.e. statically link everything, which is nice for binary distribution) -- it was a way of reducing behavioural differences between bytecode and native code outputs (something the toolchain always strived for) while retaining benefits bytecode may provide for certain things (e.g.  `ocamldebug`). 

So what's the story now ?

ghost commented:
> The toolchain consistently allows you to produce executables without extension and then you can simply add .exe if you are on Windows.

On the other hand the format is usually not determined by the extension. If you do `ocamlopt x.ml -o foo.o` then `foo.o` will still be an executable. Here it is different as the extension matters. Another option would be to force the format via another command line argument.

> However for me the idea of -custom was that it would produce the moral equivalent for bytecode of what ocamlopt does (i.e. statically link everything, which is nice for binary distribution) -- it was a way of reducing behavioural differences between bytecode and native code outputs (something the toolchain always strived for) while retaining benefits bytecode may provide for certain things (e.g. ocamldebug).

That's exactly how we have been using `-custom` in dune as well. This was making Debian sad because binaries produced by `-custom` cannot be stripped. We discussed it and we decided to extend `-output-complete-obj` to be able to produce executable as well. The binaries produced by this PR can serve the same purpose as the ones currently produced by `-custom` but in addition they can be stripped. Indeed, instead of doing a plain file concatenation of the bytecode file at the end of a custom runtime, `-output-complete-obj` produces a `.c` file with the bytecode as a constant array and compile and link this file directly into the custom runtime. This way such binaries are proper binaries that can be stripped and everyone is happy.

> So what's the story now ?

Use `-output-complete-obj` rather than `-custom`.

dbuenzli commented:
> On the other hand the format is usually not determined by the extension. If you do `ocamlopt x.ml 
-o foo.o` then `foo.o` will still be an executable. Here it is different as the extension matters.

This is why unless you use another argument to specify the format I think it's preferable if no known extension implies executable. That way you can consistently use `-o $(PROGNAME)$(EXE_EXT)` in your `Makefile` for producing your various executables. What is proposed by this PR is just very annoying.

> Use `-output-complete-obj` rather than `-custom`.

Well it doesn't really have the same properties, e.g. no `ocamldebug`.

ghost commented:
> This is why unless you use another argument to specify the format I think it's preferable if no known extension implies executable. That way you can consistently use -o $(PROGNAME)$(EXE_EXT) in your Makefile for producing your various executables. What is proposed by this PR is just very annoying.

That seems dodgy to me as you can't distinguish between something that is an unknown extension and something that is just a dot in the executable name. In particular that means that you can't detect typos.

> Well it doesn't really have the same properties, e.g. no ocamldebug.

Yes. Although, for ocamldebug specifically it wouldn't be difficult to recover it.

dbuenzli commented:
> That seems dodgy to me as you can't distinguish between something that is an unknown extension and something that is just a dot in the executable name. In particular that means that you can't detect typos.

If you are concerned about this solution then do the right thing. The toolchain allows to produce executable without `.exe` extensions and in particular `ocamlc -custom -o` does. There's really no good reason why this compilation path should introduce more idiosyncrasies than there already is in the compiler drivers.

EDIT: In this particular case the type could simply be specified as an optional argument to the `-output-complete-obj` option.

gasche commented:
Sorry if the question is obvious, but why not just use `-output-complete-exe <any-name>` to produce an executable instead of an object file?

nojb commented:
> Sorry if the question is obvious, but why not just use `-output-complete-exe <any-name>` to produce an executable instead of an object file?

Why not? Are there any objections?

dbuenzli commented:
At that rate why simply not simply add `-output-complete-exe` and reuse the `-o` we all know of ?

xavierleroy commented:
> So what's the story now ?

There are three properties for `ocamlc`-generated executables, each being desirable to a subset of OCaml users:
1. Static linking: a single, native executable contains the bytecode, the virtual machine, and C stub code.
2. Support for ocamldebug debugging and everything that can be done today with pure bytecode executables.
3. Resistance to `strip`, meaning that the generated executable should either be a `#!` script plus bytecode, or a proper native executable, but not a combination of both.

So far we have two solutions that satisfy two of these properties:
* `ocamlc` + dynamic loading of C stub code: 2 and 3
* `ocamlc -custom`: 1 and 2.

The proposal in this PR is to have a third solution that satisfies 1 and 3, as a way to please our Dune constituency and our Debian constituency.

For the record, I still think that dynamic loading of C stub code is the way to go.  Static linking is so 20th century...  So I don't care about 1 and would gladly kill `ocamlc -custom` and the proposal in this PR.  Others disagree.

gasche commented:
@dbuenzli: yes of course, let me adopt your proposal as if I had made it in the first place. (To my defense, I've essentially never used any of these options before.)

nojb commented:
OK, I changed the implementation to use the new option `-output-complete-exe` (in combination with `-o`), as suggested, and updated the manual. This is ready for review again.

ghost commented:
Thanks @nojb for this work and everybody else for their input!
